### PR TITLE
Omniauth

### PR DESCRIPTION
**Pull request by pereslop**

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -56,10 +56,10 @@ class User < ApplicationRecord
 
   scope :ordered, -> { order(username: :asc) }
 
-  def self.from_omniauth(auth, current_user)
-    authorization = Authorization.where(provider: auth.provider, uid: auth.uid.to_s).first_or_create
+  def self.from_omniauth(auth)
+    authorization = Authorization.where(provider: auth[:provider], uid: auth[:uid].to_s).first_or_create
     if authorization.user.blank?
```
Comment (Snick555) [inline app/models/user.rb]: what do you think about this?
```
return authorization.user if authorization.user
User.find_or_create_by(email: auth.info.email) do |user|
  user.password = Devise.friendly_token[0, 20]
  user.fetch_details(auth)
  user.authorizations << authorization
end
```

File: spec/controllers/users/omniauth_callbacks_controller_spec.rb [hunk 2]
```diff
@@ -1,5 +1,44 @@
 require 'rails_helper'
 
-RSpec.describe OmniauthCallbacksController, type: :controller do
+RSpec.describe Users::OmniauthCallbacksController, type: :controller do
+  let(:user) { User.find_by(email: 'pereslop@rere.com')}
```
Comment (Snick555) [inline spec/controllers/users/omniauth_callbacks_controller_spec.rb]: please add test for authorized user

File: spec/models/user_spec.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  describe 'User authorization' do
+    it 'create new authorization and user' do
+      expect do
+        described_class.from_omniauth(Faker::Omniauth.facebook)
+      end.to change(Authorization, :count).by(1).and change(described_class, :count).by(1)
+    end
+
+    it 'login registered user with authorization' do
+      authorization = FactoryGirl.create(:authorization)
+      user = authorization.user
```
Comment (Snick555) [inline spec/models/user_spec.rb]: you don't need here user variable

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+      expect do
+        described_class.from_omniauth(authorization)
+      end.to change(described_class, :count).by(0)
+    end
+
+    it 'login user without authorization' do
+      user = FactoryGirl.create(:user)
```
Comment (Snick555) [inline spec/models/user_spec.rb]: please use let for variables

File: spec/models/user_spec.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      authorization = Faker::Omniauth.facebook
```
Comment (Snick555) [inline spec/models/user_spec.rb]: in first test you have the same code, please move it to variable

File: spec/controllers/users/omniauth_callbacks_controller_spec.rb [hunk 6]
```diff
@@ -1,5 +1,44 @@
...
+      end.to change(User, :count).by(0)
+
+    end
+  end
+
+
+
+  def setup_env_for_omniauth(email = true)
+    request.env['devise.mapping'] = Devise.mappings[:user]
+    OmniAuth.config.test_mode = true
```
Comment (Snick555) [inline spec/controllers/users/omniauth_callbacks_controller_spec.rb]: shouldn't it be in configs?

File: app/models/user.rb (hunk 1 above)
Comment (pereslop) [reply app/models/user.rb]: it looks shorter, and almost without conditions :)

Comment (pereslop) [reply app/models/user.rb]: return authorization.user if authorization.user
    user = User.find_or_create_by(email: auth[:info][:email]) do |u|
      u.password = Devise.friendly_token[0, 20]
      u.fetch_details(auth)
    end
    user.authorizations << authorization
    user

**Review by Snick555 — APPROVED** [review]

**Pull request merged**